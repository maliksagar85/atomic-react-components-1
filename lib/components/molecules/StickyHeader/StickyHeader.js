import React, { useState, useRef, useEffect } from 'react';
import styled from 'styled-components';
import PropTypes from 'prop-types';
import useScroll from '../../../hooks/useScroll';

import styles, { Wrapper } from './StickyHeader.style';

const StickyHeader = ({ children, className, defaultTopHeight = 0, tag }) => {
  const Tag = tag;
  const [isSticky, setSticky] = useState(false);
  const wrapperRef = useRef(null);

  useEffect(() => {
    window.addEventListener('scroll', () => useScroll(wrapperRef, setSticky, defaultTopHeight));
    return () => {
      window.removeEventListener('scroll', useScroll);
    };
  });

  return (
    <Wrapper
      className={`${isSticky ? className : 'no-sticky'}`}
      ref={wrapperRef}
      elmRef={wrapperRef}
    >
      {tag ? (
        <Tag className="content">{children}</Tag>
      ) : (
        <header className="content">{children}</header>
      )}
    </Wrapper>
  );
};

StickyHeader.propTypes = {
  children: PropTypes.string.isRequired,
  className: PropTypes.string.isRequired,
  defaultTopHeight: PropTypes.number.isRequired,
  tag: PropTypes.string.isRequired,
};

const StyledStickyHeader = styled(StickyHeader)`
  ${styles};
`;

export default StyledStickyHeader;

export { StickyHeader as StickyHeaderVanilla };
