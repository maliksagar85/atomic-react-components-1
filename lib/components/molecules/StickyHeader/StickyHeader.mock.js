const defaultConfig = {
  variation1: {
    backgroundColor: '#ff00c8',
    padding: '5px',
    color: '#ffffff',
  },
  variation2: {
    backgroundColor: '#6e00ff',
    padding: '5px',
    color: '#ffffff',
    width: '70%',
  },
  variation3: {
    backgroundColor: '#578ca9',
    padding: '5px',
    color: '#ffffff',
    width: '350px',
  },
};

export default defaultConfig;
