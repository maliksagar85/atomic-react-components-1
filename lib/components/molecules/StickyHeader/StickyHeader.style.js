import styled, { css } from 'styled-components';

const styles = css`
  .content {
    position: fixed;
    top: ${(props) =>
      props.defaultTopHeight !== undefined ? `${props.defaultTopHeight}px` : '0px'};
    width: inherit;
  }
`;

export const Wrapper = styled.div`
  height: ${(props) => props.elmRef.current != null && `${props.elmRef.current.offsetHeight}px`};
  width: ${(props) => props.elmRef.current != null && `${props.elmRef.current.offsetWidth}px`};
`;

export default styles;
