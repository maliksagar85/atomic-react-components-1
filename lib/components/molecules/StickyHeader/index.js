export { default } from './StickyHeader';
export { StickyHeaderVanilla } from './StickyHeader';
