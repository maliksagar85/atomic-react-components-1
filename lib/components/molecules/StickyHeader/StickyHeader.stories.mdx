import { Meta, Story, Preview, Props } from '@storybook/addon-docs/blocks';

import StickyHeader, {StickyHeaderVanilla} from './StickyHeader';
import defaultConfig from './StickyHeader.mock';


<Meta title="Molecules/StickyHeader" component={StickyHeaderVanilla} />

### Knobs
<Preview>
  <Story name="Knobs">
    <div style={{ height: '1200px', overflowY: 'scroll' }}>
      <p>
        Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has
        been the industrys
      </p>
      <StickyHeaderVanilla>
        <h1 style={{ ...defaultConfig.variation1 }}>Sticky Header: No content width</h1>
      </StickyHeaderVanilla>
    </div>
  </Story>
</Preview>

### StickyHeader
<Preview>
  <Story name="StickyHeader">
    <div style={{ height: '1200px', overflowY: 'scroll' }}>
      <p>
          Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has
          been the industrys
        </p>
        <StickyHeader>
          <h1 style={{ ...defaultConfig.variation1 }}>Sticky Header: No content width</h1>
        </StickyHeader>
        <p>
          Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has
          been the industrys
        </p>
        <p>
          Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has
          been the industrys
        </p>
        <StickyHeader defaultTopHeight={34} tag="div">
          <h1 style={{ ...defaultConfig.variation2 }}>
            Sticky Header: Content width in percentage
          </h1>
        </StickyHeader>
        <p>
          Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has
          been the industrys
        </p>
        <p>
          Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has
          been the industrys
        </p>
        <StickyHeader defaultTopHeight={68} tag="nav">
          <h1 style={{ ...defaultConfig.variation3 }}>Sticky Header: Fixed content width</h1>
        </StickyHeader>
        <p>
          Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt
          ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation
          ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in
          reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur
          sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id
          est laborum.
        </p>
    </div>
  </Story>
</Preview>



## Props ðŸ”§

<Props of={StickyHeaderVanilla} />

