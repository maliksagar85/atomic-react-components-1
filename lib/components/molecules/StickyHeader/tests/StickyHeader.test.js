import React from 'react';
import { shallow, mount } from 'enzyme';
import { StickyHeaderVanilla } from '../StickyHeader';

describe('Test render <StickyHeader /> Component', () => {
  let Component;

  beforeEach(() => {
    Component = mount(<StickyHeaderVanilla />);
  });

  it('should render Header element', () => {
    expect(Component.find('header').length).toEqual(1);
  });

  it('should test children prop', () => {
    Component = mount(
      <StickyHeaderVanilla>
        <h1>Header</h1>
      </StickyHeaderVanilla>
    );

    expect(Component.props().children).toEqual(<h1>Header</h1>);
  });

  it('Should test component render correctly', () => {
    expect(Component).toMatchSnapshot();
  });
});

describe('should test component on window add or remove event', () => {
  let Component;
  const events = {};
  const setSticky = jest.fn();

  jest.spyOn(React, 'useState').mockImplementation((isSticky) => [isSticky, setSticky]);
  jest.spyOn(React, 'useEffect').mockImplementation(() => {});

  it('should test component on window addEventListener', () => {
    jest.spyOn(window, 'addEventListener').mockImplementation((event, handle) => {
      events[event] = handle;
    });

    Component = mount(<StickyHeaderVanilla />);
    events.scroll();
    console.log('1111111====', Component.find('div').html());

    expect(Component.find('div').hasClass('no-sticky')).toBeTruthy();
  });

  it('should test component on window removeEventListener', () => {
    jest.spyOn(window, 'removeEventListener').mockImplementation((event) => {
      events[event] = undefined;
    });

    Component = mount(<StickyHeaderVanilla />);
    events.scroll();
    console.log('2222222222===', Component.find('div').html());

    expect(Component.find('div').hasClass('no-sticky')).toBeTruthy();
  });

  it('should render tag from props', () => {
    Component = shallow(<StickyHeaderVanilla tag="nav" />);
    expect(Component.find('nav').length).toEqual(1);
  });

  it('should contain content class', () => {
    expect(Component.find('nav').hasClass('content')).toBeTruthy();
  });
});
