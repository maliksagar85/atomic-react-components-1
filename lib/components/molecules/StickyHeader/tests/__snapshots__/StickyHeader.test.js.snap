// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test render <StickyHeader /> Component Should test component render correctly 1`] = `
<StickyHeader>
  <StickyHeaderstyle__Wrapper
    className="no-sticky"
    elmRef={
      Object {
        "current": <div
          class="StickyHeaderstyle__Wrapper-sc-1npzz5d-0 bhETWp no-sticky"
        >
          <header
            class="content"
          />
        </div>,
      }
    }
  >
    <div
      className="StickyHeaderstyle__Wrapper-sc-1npzz5d-0 bhETWp no-sticky"
    >
      <header
        className="content"
      />
    </div>
  </StickyHeaderstyle__Wrapper>
</StickyHeader>
`;
